### PR TITLE
Feature/12 support nested matchers

### LINKED ISSUE #12: Support nested matchers
# Situations

Often result sets contain changing or even non-deterministic contents. In those cases you need more relaxed result matching than simple comparing against a constant value:

1. Structural checks
1. Syntactical checks (e.g. for a timestamp column)
1. Range checks
1. Data type checks
1. Any value

Hamcrest builds on the idea of nested matchers for exactly this particular purpose, so the `hamcrest-resultset-matcher` should support nested matchers to.

# Acceptance criteria

1. Resultset matcher supports nested Hamcrest matchers

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #12.

### CODE REVIEW

File: src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java [hunk 1]
```diff
@@ -17,26 +22,54 @@
  * The matcher supports strict type matching and a fuzzy mode. In fuzzy mode it value matches are accepted if a the
  * matcher knows how to convert between the expected type and the actual and the converted value matches.
  * </p>
+ *
+ * @param <T> Matcher type
```
Comment (jakobbraun) [inline src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java:26]: Why do you need this type. Isn't it always `ResultSet`?

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: doc/changes/changes_1.0.0.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# hamcrest-resulset-matcher 1.0.0, released 2020-10-25
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply doc/changes/changes_1.0.0.md:1]: Oopps. I am ahead of my time, it seems.

File: src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java (hunk 1 above)
Comment (redcatbear) [reply src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java:26]: You have a point. I should only support `ResultSet` and derived interfaces.

File: src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java [hunk 3]
```diff
@@ -17,26 +22,54 @@
...
+        }
+        return tableOfMatchers;
+    }
+
+    private List<Matcher<? super T>> wrapExpecteRowInMatchers(final List<Object> expectedRow) {
+        final List<Matcher<? super T>> rowOfMatchers = new ArrayList<>(expectedRow.size());
+        for (final Object expectedCellValue : expectedRow) {
+            if (expectedCellValue instanceof Matcher<?>) {
+                rowOfMatchers.add(castToMatcher(expectedCellValue));
+            } else if (this.fuzzy) {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java:61]: No. For the whole table. Not per cell.

File: src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java [hunk 4]
```diff
@@ -124,7 +121,7 @@ protected boolean matchesSafely(final ResultSet resultSet) {
         boolean ok = matchColumns(resultSet);
         try {
             int rowIndex = 0;
-            for (final List<Matcher<? super T>> expectedRow : this.expectedTable) {
+            for (final List<Matcher<?>> expectedRow : this.cellMatcherTable) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java:124]:
```suggestion
            for (final List<Matcher<?>> cellMatcherRow : this.cellMatcherTable) {
```

File: src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java [hunk 5]
```diff
@@ -170,10 +167,10 @@ private int getExpectedColumnCount() {
     }
 
     private boolean matchValuesInRowMatch(final ResultSet resultSet, final int rowIndex,
-            final List<Matcher<? super T>> expectedRow) {
+            final List<Matcher<?>> expectedRow) {
         int columnIndex = 0;
         try {
-            for (final Matcher<? super T> expectedValue : expectedRow) {
+            for (final Matcher<?> expectedValue : expectedRow) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java:173]:
```suggestion
            for (final Matcher<?> cellMatcher : expectedRow) {
```

File: src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java [hunk 6]
```diff
@@ -170,10 +167,10 @@ private int getExpectedColumnCount() {
     }
 
     private boolean matchValuesInRowMatch(final ResultSet resultSet, final int rowIndex,
-            final List<Matcher<? super T>> expectedRow) {
+            final List<Matcher<?>> expectedRow) {
```
Comment (jakobbraun) [inline src/main/java/com/exasol/matcher/ResultSetStructureMatcher.java:170]:
```suggestion
            final List<Matcher<?>> cellMatcherRow) {
```

**Review by jakobbraun — CHANGES_REQUESTED** [review]

**Review by jakobbraun — APPROVED** [review]

**Pull request merged**